src/magic_folder/client.py

```diff
@@ -0,0 +1,137 @@
+# Copyright 2020 Least Authority TFA GmbH
+# See COPYING for details.
+
+import json
+
+from twisted.internet.defer import (
+    inlineCallbacks,
+    returnValue,
+)
+from twisted.internet.error import (
+    ConnectError,
+)
+
+from twisted.web import (
+    http,
+)
+
+from hyperlink import (
+    DecodedURL,
+)
+
+from treq.client import (
+    HTTPClient,
+)
+
+import attr
+
+
+class ClientError(Exception):
+    """
+    Base class for all exceptions in this module
+    """
+
+
+class CannotAccessApiError(ClientError):
+    """
+    The Magic Folder HTTP API can't be reached at all
+    """
+
+
+@attr.s(frozen=True)
+class MagicFolderApiError(ClientError):
+    """
+    A Magic Folder HTTP API returned a failure code.
+    """
+    code = attr.ib()
+    body = attr.ib()
+
+    def __repr__(self):
+        return "<MagicFolderApiError code={} body={!r}>".format(
+            self.code,
+            self.body,
+        )
+
+    def __str__(self):
+        return u"Magic Folder HTTP API reported error {}: {}".format(
+            self.code,
+            self.body,
+        )
+
+
+@inlineCallbacks
+def _get_content_check_code(acceptable_codes, res):
+    """
+    Check that the given response's code is acceptable and read the response
+    body.
+
+    :raise MagicFolderApiError: If the response code is not acceptable.
+
+    :return Deferred[bytes]: If the response code is acceptable, a Deferred
+        which fires with the response body.
+    """
+    body = yield res.content()
+    if res.code not in acceptable_codes:
+        raise MagicFolderApiError(res.code, body)
+    returnValue(body)
+
+
+@attr.s
+class MagicFolderClient(object):
+    """
+    An object that knows how to call a particular Magic Folder HTTP API.
+
+    :ivar HTTPClient http_client: The client to use to make HTTP requests.
+
+    :ivar callable get_api_token: returns the current API token
+    """
+
+    # we only use the path-part not the domain
+    base_url = DecodedURL.from_text(u"http://invalid./")
+    http_client = attr.ib(validator=attr.validators.instance_of(HTTPClient))
+    get_api_token = attr.ib()
+
+    def list_folders(self, include_secret_information=None):
```

Comment:
Well, I wanted to be able to distinguish in case we actually wanted to send the query-arg with "=0" whereas `None` means to not include the query-arg at all. Although, there isn't a code-path that includes `=0` so I guess False or None is  kind-of equivalent here as it sits.

Also, "in general" I usually always use `None` for "I want the default" in kwargs .. probably because "sometimes" (i.e. mutable defaults) it's wrong to put the "actual" default there.